### ISSUE TITLE
Add the ability to run multiple benchmarks concurrently on same machine(s)

### ISSUE THREAD

**Issue by yangsophia** [GoogleCloudPlatform/PerfKitBenchmarker#436]
No real workloads are run on an isolated environment, being able to run more than one benchmarks on the same machine(s) at the same time could help simulate/profile more complex situation.

**Comment by yangsophia** [comment]
With this request being implemented, we can then deprecate the existing "cpu burning" change.

**Comment by voellm** [comment]
The idea is to have a flag like --benchmarks=fio+iperf

Raising pri so we pick this up sooner.

**Comment by mateusz-blaszkowski** [comment]
@voellm - just to be sure - what would be the expected behavior when passing `--benchmarks=fio+iperf`? Currently when I pass both `--benchmarks=fio,iperf` and `--parallelism=2` then these two benchmarks will be executed together but their Run phase will not be synchronized (they will actually start benchmarking at different times). Would `--benchmarks=fio+iperf` solve this issue?

**Comment by voellm** [comment]
Correct... the intent of +/is meant to mean run together on the same
machines at the same time.

**Comment by hildrum** [comment]
What's the expected machine set in this example when the configurations for `fio` and `iperf`?  I can think of several interpretations:
* two VMs, where both participate in `iperf` and one is used by the `fio` benchmark
* two VMs, where both participate in `iperf` and two copies of `fio` are run
* three VMs, where two are in `iperf` and one is running `fio`.

In this case, it might be possible to reconcile the requirements in a sensible way, but more generally I don't think that's possible.  I think there'd need to be some more configuration to determine which one of those is wanted.

Another potential problem is if the chosen benchmarks have different runtimes.   If you choose `benchmarks=A+B` and benchmark A runs for one minute and benchmark B runs for 10, the time for A would reflect the how long it takes to run A when B is running, but since A finishes much before B, you'd not get an a good idea of how long B takes to run when workload similar to A is running.

**Comment by voellm** [comment]
The second is desired.  For any single VM benchmark is should be duplicated
on all machines.

Not what happens if you try iperf+cassandra ?  That I would say is
undefined.  We should not allow mapping of configs that both need multiple
VMs.

So any one benchmark can have multiple VMs + any others that require only a
single VM.

For example iperf+fio+coremark should result in two VM's running iperf and
on each of those machines fio and coremark should be running.

On Wed, Nov 18, 2015 at 4:56 PM, Kris Hildrum <notifications@github.com>
wrote:

> What's the expected machine set in this example when the configurations
> for fio and iperf? I can think of several interpretations:
>
>    - two VMs, where both participate in iperf and one is used by the fio
>    benchmark
>    - two VMs, where both participate in iperf and two copies of fio are
>    run
>    - three VMs, where two are in iperf and one is running fio.
>
> In this case, it might be possible to reconcile the requirements in a
> sensible way, but more generally I don't think that's possible. I think
> there'd need to be some more configuration to determine which one of those
> is wanted.
>
> Another potential problem is if the chosen benchmarks have different
> runtimes. If you choose benchmarks=A+B and benchmark A runs for one
> minute and benchmark B runs for 10, the time for A would reflect the how
> long it takes to run A when B is running, but since A finishes much before
> B, you'd not get an a good idea of how long B takes to run when workload
> similar to A is running.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/issues/436#issuecomment-157912612>
> .
>



-- 
Anthony F. Voellm (aka Tony)
Google Voice:  (650) 516-7382
<https://www.google.com/voice/b/0?pli=1#phones>
Blog: http://perfguy.blogspot.com/
Benchmark like a pro... PerfKitBenchmarker
<https://github.com/GoogleCloudPlatform/PerfKitBenchmarker>

**Comment by noahl** [comment]
To solve the problem Kris pointed out about benchmarks having different runtimes, maybe we could separate the notion of a "workload" and a benchmark, and say that a "workload" is something that can run for any amount of time and always runs on a single VM. In any PKB run, you could have one benchmark (which would determine the overall runtime) and any number of workloads. You'd do something like `pkb.py --benchmarks=iperf --workloads=fio` to run the iperf+fio example.

It would still be nice to collect statistics (like throughput) from the workload, though.

**Comment by voellm** [comment]
The last thing left for this CPU is background disk activity.

**Comment by hildrum** [comment]
Memory, too, I think, unless that's been dropped off the desired list?

**Comment by voellm** [comment]
Yes memory on the list.
